### PR TITLE
GH-125174: Make immortal objects more robust, following design from PEP 683

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-125174

### CODE REVIEW

**Comment by markshannon** [comment]
This is not a complete fix for #125174, just enough to unblock other work

**Comment by markshannon** [comment]
The two "failing" refleak buildbots seem to be hanging in `test.test_concurrent_futures.test_shutdown`. There are no reported refleaks.

File: Include/refcount.h [hunk 1]
```diff
@@ -21,35 +21,48 @@ cleanup during runtime finalization.
...
 /*
-In 64+ bit systems, an object will be marked as immortal by setting all of the
+In 64+ bit systems, any object whose 32 bit reference count is >= 2**31
+will be treated as immortal.
+
+In order to offer some resilience to C extensions using the stable ABI
+compiled against 3.11 or earlier, we set the initial value near the
+middle of the range (2**31, 2**32 - 2**29). That way the
+
+will be marked as immortal by setting all of the
```
Comment (iritkatriel) [inline Include/refcount.h]: something's missing here where the empty line is

**Review by iritkatriel — APPROVED** [review]

**Comment by markshannon** [comment]
[Windows](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20241010-3.14.0a0-94f8fd0) shows an over 2% improvement. With the regression fixed, this I would expect this to be more than 3%.

This reason for this is that the MSVC does not convert `unsigned int x; x > (1 << 31)` into `signed x; x < 0` as do clang and gcc.
This PR uses the `signed x; x < 0` form. Hence the speedup.

**Comment by markshannon** [comment]
[x86-64 linux performance](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20241010-3.14.0a0-9f86e46) numbers are a bit noisy, but show no significant change.

**Pull request merged**